### PR TITLE
Data loader for seq of label model

### PR DESCRIPTION
**Pull request by fayejf**
Add feature loader to accept external features stored in pickle file.
Add feature to seq_label data loader
Add function to convert seq_label to relative labels.
Add test and test data.

Input manifest file looks like: 
```
{"feature_filepath": "tests/.data/asr/feat/AMI-00IB4001-000.p", "seq_label": "zero one two two one two one two two one two two three one two two two zero two one one two two one one two one one one one...."}
{"feature_filepath": "tests/.data/asr/feat/AMI-00IB4001-001.p", "seq_label": "0 1 0 2 1 1 0 1 1 1 1........"}
```
Output looks like:
Tensor(external feature), Tensor(feature length) ,Tensor(label), Tensor(length of seq list) 

```
(tensor([[-0.6812,  0.5182, -0.3697,  ..., -0.4531, -0.8935, -1.6601],                                                               
        [-0.2840,  0.8350,  0.9644,  ...,  0.0409,  0.0546, -0.6683],                                                                
.....), tensor(50), tensor([0., 1., 2., 2., 1., 2., 1., 2., 2., 1., 2
., 2., 3., 1., 2., 2., 2., 0.,
        2., 1., 1., 2., 2., 1., 1., 2., 1., 1., 1., 1., 0., 0., 0., 2., 0., 2.,
        2., 2., 1., .......]), tensor(50))
```

### CODE REVIEW

**Comment by nithinraok** [comment]
is it draft or ready to review?
Are there any feature models that uses these loaders, if isn't better to have it merge along with it ?!

**Comment by fayejf** [comment]
> is it draft or ready to review?
> Are there any feature models that uses these loaders, if isn't better to have it merge along with it ?!
 
@nithinraok  ready to review. Well I thought about it. Tried to split huge PR into small things. And the feature loader should be less correlated with models.

File: nemo/collections/asr/parts/collections.py
Comment (nithinraok) [inline nemo/collections/asr/parts/collections.py:341]: doc strings missing

File: nemo/collections/asr/parts/collections.py
Comment (nithinraok) [inline nemo/collections/asr/parts/collections.py:381]: what is brak?

File: nemo/collections/asr/parts/collections.py
Comment (nithinraok) [inline nemo/collections/asr/parts/collections.py:369]: why is an OR operator required here?

File: tests/collections/asr/test_label_datasets.py
Comment (nithinraok) [inline tests/collections/asr/test_label_datasets.py:28]: are these labels uniq per session or for whole manifest?
while training speaker diarizer we train per session, doesn't this mix with remaining classes and spoil the relative number while training?

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/parts/collections.py
Comment (titu1994) [inline nemo/collections/asr/parts/collections.py:381]: Break*

**Comment by nithinraok** [comment]
General questions on overall feature loader:
instead of seq label inputted through manifest should we support reading from file?
any helper function to get seq_label from rttm file?
does each row in manifest correspond to each session or part of session? if its part of session can that be handled automatically for seq_label inputs (grouth truth)

File: nemo/collections/asr/parts/collections.py
Comment (titu1994) [inline nemo/collections/asr/parts/collections.py:386]: Define what is dnc here or change name of function ?

File: nemo/collections/asr/parts/feature_loader.py
Comment (titu1994) [inline nemo/collections/asr/parts/feature_loader.py:27]: Why .p ? The standard in python is .pt for pytorch, .npy for numpy and .pkl for pickle objects. Also don't make it so brittle as to check filepath extension, there's many cases where pickle files are written with different extension.

File: nemo/collections/asr/parts/feature_loader.py
Comment (titu1994) [inline nemo/collections/asr/parts/feature_loader.py:28]: File not closed. Use with context for file and load into pickle

File: tests/collections/asr/test_label_datasets.py
Comment (titu1994) [inline tests/collections/asr/test_label_datasets.py:97]: Can you disable formatting here and just flatten this list ? Use `#fmt: off` and `#fmt: on` before and after respective to this tensor creation.

**Review by titu1994 — CHANGES_REQUESTED** [review]
Mostly ok, but needs better refactoring + needs to be less dependent on filepath extension. .p is not standard.

**Comment by fayejf** [comment]
> General questions on overall feature loader:
> instead of seq label inputted through manifest should we support reading from file?

reading from file? could you elaborate that?
> any helper function to get seq_label from rttm file?

No right now the emb and seq_label is generated offline. That is **not related to loader but data preparation imo** and shall include in the following PRs. 


> does each row in manifest correspond to each session or part of session? if its part of session can that be handled automatically for seq_label inputs (grouth truth)

Yes each row in manifest correspond to part of session.  No need to do some argumentation offline and as above said shall include them later. (After experimenting if it's worth to be included.)

File: nemo/collections/asr/parts/collections.py
Comment (fayejf) [reply nemo/collections/asr/parts/collections.py:381]: oops typo sorry

File: nemo/collections/asr/data/feature_to_label.py
Comment (fayejf) [reply nemo/collections/asr/data/feature_to_label.py:97]: unique labels. Used for future real speaker id monitoring.

File: nemo/collections/asr/data/feature_to_label.py
Comment (fayejf) [reply nemo/collections/asr/data/feature_to_label.py:70]: We expect the input feature could be other features such as mels for speech segment (maybe trained clustering model on feature directly later? probably not! just an example)  but still fall into _FeatureSeqSpeakerLabelDataset

File: nemo/collections/asr/parts/collections.py
Comment (fayejf) [reply nemo/collections/asr/parts/collections.py:386]: yes I should eliminate ambiguous dnc here.

File: nemo/collections/asr/parts/feature_loader.py
Comment (fayejf) [reply nemo/collections/asr/parts/feature_loader.py:27]: Good catch. It should be .pickle or .pkl..... .p is too confusing. 
Sure will have a smarter way to check and open pickle file.

File: nemo/collections/asr/parts/feature_loader.py
Comment (fayejf) [reply nemo/collections/asr/parts/feature_loader.py:28]: good catch

File: nemo/collections/asr/parts/feature_loader.py
Comment (fayejf) [reply nemo/collections/asr/parts/feature_loader.py:46]: Good question.  Follow the way as in ASR segment. And there could be some augmentor that can not work on torch tensor I think, so is safer to keep it this way and leave it to augmentor

File: nemo/collections/asr/parts/feature_loader.py
Comment (fayejf) [reply nemo/collections/asr/parts/feature_loader.py:21]: Followed the way in asr audio loader that could load as int32 instead of float32. Yes not implemented and useful here. Will remove for now

File: nemo/collections/asr/data/feature_to_label.py
Comment (fayejf) [reply nemo/collections/asr/data/feature_to_label.py:87]: We expect the input feature could be other features such as mels for speech segment (maybe trained clustering model on feature directly later? probably NOT! just an example) but still fall into _FeatureSeqSpeakerLabelDataset. And allow the user to use their magic feature for the seq_speaker_label task

File: nemo/collections/asr/parts/feature_loader.py
Comment (fayejf) [reply nemo/collections/asr/parts/feature_loader.py:27]: ok `np.load(..., allow_pickle=True)` is a good way to take care of `pkl, npz, and npy`  at the same time!

File: nemo/collections/asr/parts/collections.py
Comment (fayejf) [reply nemo/collections/asr/parts/collections.py:369]: Just to join two sets.

File: nemo/collections/asr/parts/feature_loader.py
Comment (fayejf) [reply nemo/collections/asr/parts/feature_loader.py:22]: No this should be file_path for feature file here not manifest_filepath

File: tests/collections/asr/test_label_datasets.py
Comment (fayejf) [reply tests/collections/asr/test_label_datasets.py:28]: It's fine. We use the parse to convert to relative speakers for each seq_label

File: nemo/collections/asr/parts/feature_loader.py
Comment (fayejf) [reply nemo/collections/asr/parts/feature_loader.py:22]: No this is not manifest_filepath. It's feature (pickle) file path

**Comment by fayejf** [comment]
@nithinraok 
> instead of seq label inputted through manifest should we support reading from file?

Not for now. It will make data augmentation much complicated.
> any helper function to get seq_label from rttm file?

Yes it will be provided in following PR. out of scope here

> does each row in manifest correspond to each session or part of session? if its part of session can that be handled automatically for seq_label inputs (grouth truth)

Could be each or part. Not quite understand this "if its part of session can that be handled automatically for seq_label inputs (grouth truth)". If I understand your question correctly, theoretically yes but as in above question, the augmentation will be difficult. This will be handled in neural clustering in clustering diarizer.

File: nemo/collections/asr/parts/feature_loader.py
Comment (nithinraok) [reply nemo/collections/asr/parts/feature_loader.py:20]: Add doc strings here

**Review by nithinraok — APPROVED** [review]
LGTM

File: nemo/collections/asr/data/feature_to_label.py
Comment (titu1994) [inline nemo/collections/asr/data/feature_to_label.py:25]: Add docstring for pad_id

File: nemo/collections/asr/data/feature_to_label.py
Comment (titu1994) [reply nemo/collections/asr/data/feature_to_label.py:71]: Base classes shouldnt be handling special cases like this, but for now lets leave it as is.

File: nemo/collections/asr/parts/collections.py
Comment (titu1994) [inline nemo/collections/asr/parts/collections.py:340]: Write full form, not acronyms here unless necessary

File: nemo/collections/asr/parts/collections.py
Comment (titu1994) [inline nemo/collections/asr/parts/collections.py:422]: Same comment as above

File: nemo/collections/asr/parts/collections.py
Comment (titu1994) [inline nemo/collections/asr/parts/collections.py:425]: No *args and **kwargs. Repeat the kwargs from above.

File: nemo/collections/asr/parts/collections.py
Comment (titu1994) [inline nemo/collections/asr/parts/collections.py:432]: Repeat args and kwargs

File: nemo/collections/asr/parts/collections.py
Comment (titu1994) [inline nemo/collections/asr/parts/collections.py:442]: Why __parse_item? Just use one _ ?

File: nemo/collections/asr/parts/feature_loader.py
Comment (titu1994) [reply nemo/collections/asr/parts/feature_loader.py:20]: Add docstring

**Review by titu1994 — CHANGES_REQUESTED** [review]
Minor changes, but overall looks good

File: nemo/collections/asr/data/feature_to_label.py
Comment (fayejf) [reply nemo/collections/asr/data/feature_to_label.py:25]: The pad_id here is not used/useful. Remove it for now, will add back and implement padding part in the future if necesarry.

File: nemo/collections/asr/parts/collections.py
Comment (fayejf) [reply nemo/collections/asr/parts/collections.py:425]: Should we try to be consistent with `AudioText` and `ASRAudioText`?

File: nemo/collections/asr/parts/collections.py
Comment (fayejf) [reply nemo/collections/asr/parts/collections.py:442]: Try to be consistent with all other classes in collections.py

File: nemo/collections/asr/parts/feature_loader.py
Comment (titu1994) [reply nemo/collections/asr/parts/feature_loader.py:24]: @fayejf ^ Docstrings explaining what the filepath is and what data it should point to.

**Review by titu1994 — APPROVED** [review]
Add the docstring, LGTM otherwise

**Comment by fayejf** [comment]
> Add the docstring, LGTM otherwise

@titu1994 Thanks! Added. will merge once the checks finished.

**Pull request merged**
